candle-nn/src/encoding.rs

```diff
@@ -0,0 +1,379 @@
+///
+/// As an example, the following tensor will be encoded to a one-hot matrix:
+///
+/// ```
+/// [[0., 2.], [1., -1.]]
+/// ```
+///
+/// with a depth of 4 will be encoded to:
+///
+/// ```
+/// [[[1., 0., 0., 0.], [0., 0., 1., 0.]], [[0., 1., 0., 0.], [0., 0., 0., 0.]]]
+/// ```
+///
+/// When the input tensor index has a value of -1, the corresponding one-hot vector will be ignored,
+/// resulting in a vector of values set to the `off_value`.
+///
+///
+/// This method supports one-cold encoding by setting `on_value` to `0.0` and `off_value` to `1.0`.
+/// By default `on_value` is `1.0` and `off_value` is `0.0`.
+///
+/// Other encoding values can be used by setting `on_value` and `off_value` to the desired values.
+///
+/// # Examples
+///
+/// ## One-hot encoding
+///
+/// ```rust
+/// use candle::{Shape, Tensor, Device};
+/// use candle_nn::encoding::one_hot;
+///
+/// let device = candle::Device::Cpu;
+///
+/// let indices = Tensor::new(vec![vec![0f32, 2.], vec![1., -1.]], &device)?;
+/// let depth = 4;
+///
+/// let allow_f64 = true;
+/// let on_value = None; // defaults to 1.0
+/// let off_value = None; // defaults to 0.0
+///
+/// let one_hot = one_hot::<f32, f32>(indices, depth, allow_f64, on_value, off_value)?;
+///
+/// let expected_matrix = [
+///     [[1.0, 0.0, 0.0, 0.0], [0.0, 0.0, 1.0, 0.0]],
+///     [[0.0, 1.0, 0.0, 0.0], [0.0, 0.0, 0.0, 0.0]],
+/// ];
+///
+/// assert_eq!(one_hot.shape(), &Shape::from((2, 2, depth)));
+///
+/// let matrix = one_hot.to_vec3::<f32>()?;
+///
+/// assert_eq!(matrix, expected_matrix);
+///```
+/// ## One-cold Encoding
+///
+/// ```rust
+/// use candle::{Shape, Tensor, Device};
+/// use candle_nn::encoding::one_hot;
+///
+///
+/// let device = candle::Device::Cpu;
+/// let depth = 4;
+/// let indices = Tensor::new(vec![vec![0i64, 2], vec![1, 3]], &device)?;
+///
+/// let allow_f64 = false;
+/// let on_value = Some(0);
+/// let off_value = Some(1);
+///
+/// let one_cold = one_hot::<i64, u8>(indices, depth, allow_f64, on_value, off_value)?;
+///
+/// let expected_matrix = [[[0, 1, 1, 1], [1, 1, 0, 1]], [[1, 0, 1, 1], [1, 1, 1, 0]]];
+///
+/// assert_eq!(one_cold.shape(), &Shape::from((2, 2, depth)));
+///
+/// let matrix = one_cold.to_vec3::<u8>()?;
+///
+/// assert_eq!(matrix, expected_matrix);
+/// ```
+///
+///
+/// # Bails
+///
+/// This method will bail on tensors with a rank greater than 3.
+///
+/// This method will bail if the index value is less than -1.
+///
+/// This method will bail if the index value is greater than or equal to the depth value.
+///
+/// This method will bail if the input data type does not equal the generic data type provided.
+///
+/// This method will bail if the input data type is `f64` and `allow_f64` is false.
+///
+/// # API Design
+///
+/// The api design for this method is loosely based on the [TensorFlow One-Hot](https://www.tensorflow.org/api_docs/python/tf/one_hot) method.
+pub fn one_hot<I: WithDType, O: WithDType>(
```

Comment:
Avoid the template with types in the external facing code, at least for the input dtype here we would rather handle internall the 3 cases `u8`, `u32`, and `i64` and bail out on everything else. Same as what we're currently doing on embeddings.